From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cem Mergenci <cmergenci@gmail.com>
Date: Mon, 23 Dec 2024 22:36:57 +0300
Subject: [PATCH] Guard against panic in cty.Value.LengthInt().

Signed-off-by: Cem Mergenci <cmergenci@gmail.com>

diff --git a/internal/tags/key_value_tags.go b/internal/tags/key_value_tags.go
index 566b4ab6e8..0313e94bdd 100644
--- a/internal/tags/key_value_tags.go
+++ b/internal/tags/key_value_tags.go
@@ -797,6 +797,10 @@ func GetAnyAttr(value cty.Value, attr string, shouldReturnSetElement func(string
 			return cty.NilVal, fmt.Errorf("invalid index: %s", indexStr)
 		}
 
+		if value.Type().IsListType() && (value.IsNull() || !value.IsKnown()) {
+			return cty.NilVal, fmt.Errorf("list attribute %s is unknown or null", attrName)
+		}
+
 		if index >= value.LengthInt() {
 			return cty.NilVal, fmt.Errorf("index %d out of range for attribute %s", index, attrName)
 		}
